app/views/QRScan.js

```diff
@@ -0,0 +1,291 @@
+import React, { Component } from 'react';
+import {
+  BackHandler,
+  Dimensions,
+  ImageBackground,
+  StatusBar,
+  StyleSheet,
+  View,
+} from 'react-native';
+import { RNCamera } from 'react-native-camera';
+import QRCodeScanner from 'react-native-qrcode-scanner';
+import { SvgXml } from 'react-native-svg';
+
+import BackgroundImage from './../assets/images/launchScreenBackground.png';
+import StateAtRisk from './../assets/svgs/stateAtRisk';
+import StateNoContact from './../assets/svgs/stateNoContact';
+import { isPlatformAndroid } from './../Util';
+import { Button } from '../components/Button';
+import { Typography } from '../components/Typography';
+import Colors from '../constants/colors';
+import fontFamily from '../constants/fonts';
+import languages from '../locales/languages'; // TODO internationalization
+import LocationServices from '../services/LocationService';
+
+const StateEnum = {
+  DEFAULT: 0,
+  SCAN_SUCCESS: 1,
+  SCAN_FAIL: 2,
+  SCAN_IN_PROGRESS: 3,
+};
+
+const StateIcon = ({ status, size }) => {
+  let icon;
+  switch (status) {
+    case StateEnum.DEFAULT:
+      icon = null;
+      break;
+    case StateEnum.SCAN_SUCCESS:
+      icon = StateNoContact;
+      break;
+    case StateEnum.SCAN_FAIL:
+      icon = StateAtRisk;
+      break;
+    case StateEnum.SCAN_IN_PROGRESS:
+      icon = StateNoContact;
+      break;
+  }
+  return (
+    <SvgXml xml={icon} width={size ? size : 80} height={size ? size : 80} />
+  );
+};
+
+const height = Dimensions.get('window').height;
+
+class QRScan extends Component {
+  constructor(props) {
+    super(props);
+
+    if (isPlatformAndroid()) {
+      StatusBar.setBackgroundColor(Colors.TRANSPARENT);
+      StatusBar.setBarStyle('light-content');
+      StatusBar.setTranslucent(true);
+    }
+
+    this.state = {
+      currentState: StateEnum.DEFAULT,
+    };
+  }
+
+  componentDidMount() {
+    BackHandler.addEventListener('hardwareBackPress', this.handleBackPress);
+    const handleStateChange = () => {
+      if (this.props.navigation.isFocused()) {
+        this.onNavigate();
+      } else {
+        this.setState({ currentState: StateEnum.DEFAULT });
+      }
+    };
+    this.unsubscribeFocus = this.props.navigation.addListener(
+      'focus',
+      handleStateChange,
+    );
+    this.unsubscribeState = this.props.navigation.addListener(
+      'state',
+      handleStateChange,
+    );
+  }
+
+  componentWillUnmount() {
+    BackHandler.removeEventListener('hardwareBackPress', this.handleBackPress);
+    this.unsubscribeFocus();
+    this.unsubscribeState();
+  }
+
+  onNavigate() {
+    if (this.props.route && this.props.route.params) {
+      const { latitude, longitude } = this.props.route.params;
+      if (typeof latitude !== 'undefined' && typeof longitude !== 'undefined') {
+        this.saveCoordinates(latitude, longitude);
+      } else {
+        this.setState({ currentState: StateEnum.SCAN_IN_PROGRESS });
```

Comment:
Yes, this is just checking whether or not latitude and longitude are part of the route params. They are in the route params if and only if this screen has been brought into focus through deep linking. Otherwise, it means we got here because the user is trying to scan through the app, so we set to `SCAN_IN_PROGRESS` which renders the QR code scanner.

Reply:
I've now refactored to hopefully make this clearer:

```js
  const onFocus = () => {
    const { latitude, longitude } = getLatLonFromRouteParams();
    const isDeepLink = typeof latitude !== 'undefined' && typeof longitude !== 'undefined'
    if (isDeepLink) {
      const savedSuccessfully = saveCoordinates(latitude, longitude);
      if (savedSuccessfully) {
        setcurrentState(StateEnum.SCAN_SUCCESS);
      } else {
        setcurrentState(StateEnum.SCAN_FAIL);
      }
    } else {
      setcurrentState(StateEnum.SCAN_IN_PROGRESS);
    }
  };
```

Reply:
Cool, that makes sense and looks like it should always be in the right state. 👍